File: go/heap/min_kth_element.go
package main

import (
	"container/heap"
	"sort"
)

//最小的k的个数

//排序，取前k个
func getLeastNumbers1(arr []int, k int) []int {
	sort.Ints(arr)
	return arr[:k]
}

//使用一个堆
func getLeastNumbers2(arr []int, k int) []int {
	h := &PriorityQueue{[]int{}}
	heap.Init(h)
	for _, v := range arr {
		heap.Push(h, v)
	}

	var res []int
	for i := 0; i < k; i++ {
		res = append(res, heap.Pop(h).(int))
	}
	return res
}

type PriorityQueue struct {
	items []int
}

func (h *PriorityQueue) Len() int           { return len(h.items) }
func (h *PriorityQueue) Less(i, j int) bool { return h.items[i] < h.items[j] }
func (h *PriorityQueue) Swap(i, j int)      { h.items[i], h.items[j] = h.items[j], h.items[i] }
func (h *PriorityQueue) Push(v interface{}) { h.items = append(h.items, v.(int)) }
func (h *PriorityQueue) Pop() interface{} {
	n := len(h.items)
	x := h.items[n-1]
	h.items = h.items[:n-1]
	return x
}

File: java/datastructure/heap/leetcode/MinKthElement.java
package datastructure.heap.leetcode;

import java.util.Arrays;
import java.util.Collections;
import java.util.PriorityQueue;
import java.util.Queue;

/**
 * @author roseduan
 * @time 2020/9/16 8:54 下午
 * @description 最小的K个元素
 */
public class MinKthElement {

    /**
     * 第一种解法，排序，取前k个
     */
    public int[] getLeastNumbers1(int[] arr, int k) {
        Arrays.sort(arr);
        return Arrays.copyOf(arr, k);
    }

    /**
     * 第二种解法，使用一个优先级队列
     */
    public int[] getLeastNumbers2(int[] arr, int k) {
        Queue<Integer> queue = new PriorityQueue<>(Collections.reverseOrder());
        for (int i = 0; i < arr.length; i++) {
            if (i < k) {
                queue.add(arr[i]);
            }else {
                if (!queue.isEmpty() && arr[i] < queue.peek()) {
                    queue.poll();
                    queue.add(arr[i]);
                }
            }
        }
        int[] res = new int[k];
        int i = 0;
        for (int n : queue) {
            res[i++] = n;
        }
        return res;
    }

    /**
     * 第三种解法，还是使用一个优先级队列，只不过代码实现更加简单
     */
    public int[] getLeastNumbers3(int[] arr, int k) {
        Queue<Integer> queue = new PriorityQueue<>();
        for (int n : arr) {
            queue.add(n);
        }

        int[] res = new int[k];
        for (int i = 0; i < k && !queue.isEmpty(); i++) {
            res[i] = queue.poll();
        }
        return res;
    }

    /**
     * 第四种解法，使用快排的思想
     */
    public int[] getLeastNumbers4(int[] arr, int k) {
        return null;
    }
}
